File: Advance/10.Implement depth-first search (DFS) for a graph/Solution.py
from collections import defaultdict

class Graph:
    def __init__(self):
        self.graph = defaultdict(list)

    def add_edge(self, u, v):
        self.graph[u].append(v)
        self.graph[v].append(u)  # For undirected graph

    def dfs(self, v, visited):
        print(v, end=" ")
        visited.add(v)

        for u in self.graph[v]:
            if u not in visited:
                self.dfs(u, visited)

if __name__ == "__main__":
    g = Graph()

    n = int(input("Enter the number of nodes: "))
    edges = int(input("Enter the number of edges: "))

    print("Enter the edges (u v):")
    for _ in range(edges):
        u, v = map(int, input().split())
        g.add_edge(u, v)

    start = int(input("Enter the starting node: "))

    print("DFS Traversal: ", end="")
    g.dfs(start, set())

File: Advance/10.Implement depth-first search (DFS) for a graph/Solution.java
import java.util.*;

public class DFSGraph {
    private Map<Integer, List<Integer>> graph = new HashMap<>();
    private Set<Integer> visited = new HashSet<>();

    public void addEdge(int u, int v) {
        graph.computeIfAbsent(u, k -> new ArrayList<>()).add(v);
        graph.computeIfAbsent(v, k -> new ArrayList<>()).add(u); // For undirected graph
    }

    public void DFS(int v) {
        System.out.print(v + " ");
        visited.add(v);

        for (int u : graph.getOrDefault(v, new ArrayList<>())) {
            if (!visited.contains(u)) {
                DFS(u);
            }
        }
    }

    public static void main(String[] args) {
        DFSGraph dfsGraph = new DFSGraph();
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter the number of nodes: ");
        int n = scanner.nextInt();

        System.out.print("Enter the number of edges: ");
        int edges = scanner.nextInt();

        System.out.println("Enter the edges (u v):");
        for (int i = 0; i < edges; i++) {
            int u = scanner.nextInt();
            int v = scanner.nextInt();
            dfsGraph.addEdge(u, v);
        }

        System.out.print("Enter the starting node: ");
        int start = scanner.nextInt();

        System.out.println("DFS Traversal: ");
        dfsGraph.DFS(start);
    }
}
